File: core/src/com/inkostilation/pong/engine/Paddle.java
package com.inkostilation.pong.engine;

import com.inkostilation.pong.engine.geometry.Rectangle;

public class Paddle extends Rectangle implements IUpdatable {

    private static final float FRICTION = 0.07f;
    private static final float ACCELERATION = 40f;
    private static final float MAX_VELOCITY = 90f;

    private float yVel, defaultX, defaultY;

    private PlayerRole playerRole;

    private boolean controlled;

    private Direction accelerationDirection;

    public Paddle(float x, float y) {
        super(x, y, 20, 80);
        defaultX = x;
        defaultY = y;
        setDefaultState();
    }

    public void reset() {
        setX(defaultX);
        setY(defaultY);
        yVel = 0;
        setDefaultState();
    }

    private void setDefaultState() {
        this.yVel = 0;
        accelerationDirection = Direction.IDLE;
        this.controlled = false;
    }

    public void setAccelerationDirection(Direction accelerationDirection) {
        this.accelerationDirection = accelerationDirection;
    }

    public PlayerRole getPlayerRole() {
        return playerRole;
    }

    public void setPlayerRole(PlayerRole playerRole) { this.playerRole = playerRole; }

    public void setControlled(boolean controlled) {
        this.controlled = controlled;
    }

    public boolean isControlled() {
        return controlled;
    }

    @Override
    public void update(float delta) {
        int dir = accelerationDirection.value;
        if (dir != 0)
            yVel += ACCELERATION * dir * delta;
        else
            yVel *= (1 - FRICTION * delta);

        if (yVel >= MAX_VELOCITY)
            yVel = MAX_VELOCITY;
        else if (yVel <= -MAX_VELOCITY)
            yVel = -MAX_VELOCITY;

        setY(getY() + yVel * delta);
    }

    public void constrain(Rectangle rectangle)
    {
        if (getY() < 0) {
            setY(0);
        }
        else if (getY() > rectangle.getHeight() - getHeight()) {
            setY(rectangle.getHeight() - getHeight());
        }
    }
}

File: core/src/com/inkostilation/pong/engine/Paddle.kt
package com.inkostilation.pong.engine

import com.inkostilation.pong.engine.geometry.Rectangle

class Paddle(private val defaultX: Float, private val defaultY: Float) : Rectangle(defaultX, defaultY, 20f, 80f), IUpdatable {
    private var yVel = 0f
    var playerRole: PlayerRole? = null
    var isControlled = false
    private var accelerationDirection: Direction? = null
    fun reset() {
        x = defaultX
        y = defaultY
        yVel = 0f
        setDefaultState()
    }

    private fun setDefaultState() {
        yVel = 0f
        accelerationDirection = Direction.IDLE
        isControlled = false
    }

    fun setAccelerationDirection(accelerationDirection: Direction?) {
        this.accelerationDirection = accelerationDirection
    }

    override fun update(delta: Float) {
        val dir = accelerationDirection!!.value
        if (dir != 0) yVel += ACCELERATION * dir * delta else yVel *= 1 - FRICTION * delta
        if (yVel >= MAX_VELOCITY) yVel = MAX_VELOCITY else if (yVel <= -MAX_VELOCITY) yVel = -MAX_VELOCITY
        y = y + yVel * delta
    }

    fun constrain(rectangle: Rectangle) {
        if (y < 0) {
            y = 0f
        } else if (y > rectangle.height - height) {
            y = rectangle.height - height
        }
    }

    companion object {
        private const val FRICTION = 0.07f
        private const val ACCELERATION = 40f
        private const val MAX_VELOCITY = 90f
    }

    init {
        setDefaultState()
    }
}